File: home/.config/hypr/scripts/switcher.pl
# by vincentaxhe from China
#!/bin/perl
use strict;
use utf8;
use JSON;
use Getopt::Std;
my $usage = "Usage: $0 -[fcm]";
getopts("fcm", \my %options) or die $usage;
my %arg = reverse %options or die $usage;
my $mode = $arg{1};
my $clients_json = qx(hyprctl -j clients);
my $activeworkspace_json = qx(hyprctl -j activeworkspace);
my $clients = decode_json $clients_json;
my $activeworkspace = decode_json $activeworkspace_json;
my $workspacename = $activeworkspace->{name};
my @sortedwindows = sort {$a->{focusHistoryID} <=> $b->{focusHistoryID}} @$clients;
my (@addressarr, @menuarr);
my $i = 0;
foreach my $window (@sortedwindows){
	if ($window->{title}){
		push @addressarr, $window->{address};
		push @menuarr, join "\t", $i++, "workspace：".$window->{workspace}{name}, $window->{title};
	}
}
my $menu = join "\n", @menuarr;
my %promt = (f => "focuswindow", c => "close window", m => "movetoworkspace $workspacename");
my $out = qx(echo "$menu"|wofi --dmenu -W 40% -O alphabetical -p "$promt{$mode}") or die;
my $index = (split /\t/, $out)[0];
my $dest = $addressarr[$index] or die;
my %hyprdispatchs = (
	f => ["hyprctl", "dispatch", "focuswindow", "address:$dest"],
	c => ["hyprctl", "dispatch", "closewindow", "address:$dest"],
	m => ["hyprctl", "--batch", "dispatch focuswindow address:$dest; dispatch movetoworkspace $workspacename"]
);
exec $hyprdispatchs{$mode}->@*;

File: home/.config/hypr/scripts/switcher.py
#!/usr/bin/env python3

import gi
import sys
import os
import json
import re
import array as arr

gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

def resolveIconPath(iconName):
    iconTheme = Gtk.IconTheme.get_default()
    iconFile = iconTheme.lookup_icon(iconName.lower(), 32, 0)
    if iconFile:
        return iconFile.get_filename()
    else:
        return ""

def mapWindow(w):
    return "img:%s:text:%s"%(resolveIconPath(w["class"]), "%s (%s_%s)"%(w["title"], w["address"], w["workspace"]["id"]))

windows = json.loads(os.popen("hyprctl -j clients").read())
filtered_windows = list(filter(lambda w: w["workspace"]["id"] != -1, windows))
mapped_windows = list(map(mapWindow, filtered_windows))

print(mapped_windows)

selected_window = os.popen("echo \"%s\" | wofi -I -S dmenu"%("\n".join(mapped_windows))).read()

print("selected_window: %s"%(selected_window))

if (selected_window):
    match = re.search(r"\((\w+)\)$", selected_window)
    addr = match.group(1).split("_")[0]
    os.system("hyprctl dispatch focuswindow address:%s"%(addr))
else:
    print("no selected_window")